src/mpid/ch4/generic/mpidig_send.h

```diff
@@ -46,7 +46,15 @@ static inline int MPIDIG_isend_impl(const void *buf, MPI_Aint count, MPI_Datatyp
     MPIDIG_REQUEST(sreq, count) = count;
 #endif
 
-#ifdef MPIDI_CH4_DIRECT_NETMOD
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+    if (MPIDI_av_is_local(addr)) {
+        mpi_errno = MPIDI_SHM_am_isend(rank, comm, MPIDIG_SEND,
+                                       &am_hdr, sizeof(am_hdr), buf, count, datatype, sreq);
+        MPIR_ERR_CHECK(mpi_errno);
+
+        goto fn_exit;
+    }
+#endif
```

Comment:
In the new AM scheme, the transport is "atomic" or opaque from protocol point of view. This differentiates the shm transport from netmod transport and further tangles with protocols.

Reply:
This differentiation is already here, no? I don't think this change is furthering any entanglement.

Reply:
Before the blob of 
```
#ifdef MPIDI_CH4_DIRECT_NETMOD
MPIDI_NM_am_isend(...)
#else
if (is_local)
    MPIDI_SHM_am_isend(...)
else
    MPIDI_NM_am_isend(...)
#endif
```
could be (semantically) swapped with `MPIDIG_AM_SEND(...)`. That is, the details of how to transport is not part of am protocol.

This change made that swap no longer possible.

Reply:
I'm sorry I just don't understand what you are saying here.

Reply:
Are you referring to the rndv changes from a previous commit? Or just the preprocessor/branch rearrangement in this one?